Device and torch._C function cleanup

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #38211 Move all torch.nn.modules type annotations inline
* #39483 Upgrade lint.
* **#38173 Device and torch._C function cleanup**

- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

ezyang committed:
Device and torch._C function cleanup

- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Device and torch._C function cleanup"


- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Device and torch._C function cleanup"


- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

[ghstack-poisoned]

rgommers commented on torch/_C/__init__.pyi.in:
```diff
@@ -17,16 +19,17 @@ T = TypeVar('T')
+    # THPDevice_pynew
     @overload
-    def __init__(self, device: Union[_int, str]) -> None: ...
+    def __init__(self, device: Device) -> None: ...
```
`Device` includes `None`, which works when `device` is a keyword in other functions, but here `torch.device(None)` actually gives a `TypeError`. So I think this change should be reverted.

rgommers commented on torch/_C/__init__.pyi.in:
```diff
@@ -82,22 +85,46 @@ def _jit_get_operation(op_name: str) -> Callable: ...
+def _set_cudnn_benchmark(arg: _bool) -> None: ...  # THPModule_setBenchmarkCuDNN
+def _get_cudnn_deterministic() -> _bool: ...  # THPModule_deterministicCuDNN
+def _set_cudnn_deterministic(arg: _bool) -> None: ...  # THPModule_setDeterministicCuDNN
+# TODO: Is there a Capsule type in typing?  Doesn't seem so...
```
Nope, doesn't exist. One discussion on python-dev where the suggestion is to add it to `types`, but wasn't done. From https://code.activestate.com/lists/python-dev/139675/:

>> - capsules
>> That one seems reasonable. But maybe there should just be a `types.Capsule` Type or `types.PyCapsule`, and then you can just check that the same as any other concrete type?

> But how often do you need to verify that something is a capsule, without knowing that it's the *right* capsule? At runtime, you'd usually use `PyCapsule_IsValid`, not `PyCapsule_CheckExacf`, right? So should the type checker be tracking the name too?

rgommers reviewed (commented):
LGTM modulo my one comment on `device.__init__`.

clang-tidy failures can be ignored.

ezyang committed:
Update on "Device and torch._C function cleanup"


- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

[ghstack-poisoned]

ezyang committed:
Update on "Device and torch._C function cleanup"


- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

[ghstack-poisoned]

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -533,7 +513,6 @@ static PyMethodDef TorchMethods[] = {
   {"_add_docstr",     (PyCFunction)THPModule_addDocStr,       METH_VARARGS, nullptr},
   {"_init_names",     (PyCFunction)THPModule_initNames,       METH_O,       nullptr},
   {"_has_distributed",(PyCFunction)THPModule_hasDistributed,  METH_NOARGS,  nullptr},
-  {"_safe_call",      (PyCFunction)(void(*)())THPModule_safeCall, METH_VARARGS | METH_KEYWORDS, nullptr},
```
This is just legacy from our old `ffi` module and we hope no one uses it outside?

ezyang replied:
Yep

albanD reviewed (approved):
LGTM
The corresponding cpp functions is a nice touch as it makes looking for them easy. Is mypy actually using these comments?

ezyang committed:
Update on "Device and torch._C function cleanup"


- Introduce torch.types.Device representing all "device-like" types
- Stubbed torch.device.__reduce__
- Stubbed all torch._C functions comprehensively
- Deleted _safe_call which is unused throughout the codebase

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D21497399](https://our.internmc.facebook.com/intern/diff/D21497399)

[ghstack-poisoned]

ezyang commented:
> The corresponding cpp functions is a nice touch as it makes looking for them easy. Is mypy actually using these comments?

No, I just added them to make it more clear how to determine the correct annotations
